### PR TITLE
comm: re-implement dynamic processes using mpir-layer lpid

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description
Dependency PR: #7235 , #7237, #7242

Now both ch3 and ch4 can directly use `comm->local_group` and `comm->remote_group` (intercomm) to set up communicator and look up av addresses, we can remove the redundant code -

* revamp `ch4_spawn` to use a temporary dynamic av to exchange info between group leaders and establish intercomm
* remove ch4 `MPIDI_rank_map_t`
* remove MPIR comm mapper

[skip warnings]

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
@raffenet Let me know if you would like me to split this PR.

**Comment by raffenet** [comment]
> @raffenet Let me know if you would like me to try split this PR.

I've scanned through once and think maybe split after the ADI changes. I think what I'm seeing is that for space optimized process mappings, this PR supports dup and strided. Maybe offset? It drops support for block strided mappings at least. Any others?

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch4/most

File: src/mpid/ch4/netmod/ofi/ofi_spawn.c [hunk 1]
```diff
@@ -318,6 +318,10 @@ int MPIDI_OFI_insert_upid(MPIR_Lpid lpid, const char *upid, int upid_len)
         MPIR_Assert(MPIDI_OFI_AV_ADDR_ROOT(av) != FI_ADDR_NOTAVAIL);
     }
 
+    if (MPIDI_OFI_AV_ADDR_ROOT(av) == 0) {
+        MPIDI_OFI_global.lpid0 = lpid;
+    }
```
Comment (raffenet) [inline src/mpid/ch4/netmod/ofi/ofi_spawn.c:323]: do we always want to set this even when we aren't actually inserting into the av?

File: src/mpid/ch4/src/ch4i_comm.c [hunk 2]
```diff
@@ -92,13 +92,13 @@ int MPIDI_check_disjoint_lpids(MPIR_Lpid lpids1[], int n1, MPIR_Lpid lpids2[], i
...
         int bit = n % 32;
-        if (gpidmask[idx] & (1 << bit)) {
+        if (gpidmask[idx] & (1ULL << bit)) {
             MPIR_ERR_SET1(mpi_errno, MPI_ERR_COMM,
                           "**dupprocesses", "**dupprocesses %d", (int) lpids2[i]);
             goto fn_fail;
         }
         /* Add a check on duplicates *within* group 2 */
-        gpidmask[idx] = gpidmask[idx] | (1 << bit);
+        gpidmask[idx] = gpidmask[idx] | (1ULL << bit);
```
Comment (raffenet) [inline src/mpid/ch4/src/ch4i_comm.c:101]: commit message for `ch4/ucx: fixups in ucx_spawn.c` needs update

**Review by raffenet — COMMENTED** [review]
Biggest concern is the dynamic data exchange.

File: src/mpid/ch4/src/ch4_comm.c [hunk 3]
```diff
@@ -406,277 +406,382 @@ int MPID_Comm_set_hints(MPIR_Comm * comm_ptr, MPIR_Info * info_ptr)
...
+    /* layout:
+     *    local_size
+     *    context_id
+     *    lpids[local_size]
+     *    num_worlds
+     *    namespace[num_worlds][MPIR_NAMESPACE_MAX]
+     *    world_sizes[num_worlds]
+     *    upid_sizes[local_size]
+     *    upids[]
+     */
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/mpid/ch4/src/ch4_comm.c:518]: The data type safety is maintained by the consistency between `prepare_local_data` and `extract_remote_data`, both static local functions. The maintenance require careful matching between the two functions. Splitting into 8 or so separate dynamic send/recv will still require careful matching for each messages.  Why would the type safety be any different? It is just exchanging pointer arithmetic with "count + datatype". Most of the data need be in `MPI_Byte` anyway. so I don't think one is easier to maintain vs another. In addition, multiple messages involves multiple buffer management and it will be difficult to factor the code as the buffers are managed in `leader_exchange`.

File: src/mpid/ch4/include/mpidpre.h [hunk 4]
```diff
@@ -561,69 +561,6 @@ typedef struct MPIDIG_comm_t {
...
-    MPIDI_RANK_MAP_OFFSET,
-    MPIDI_RANK_MAP_OFFSET_INTRA,
-    MPIDI_RANK_MAP_STRIDE,
-    MPIDI_RANK_MAP_STRIDE_INTRA,
-    MPIDI_RANK_MAP_STRIDE_BLOCK,
-    MPIDI_RANK_MAP_STRIDE_BLOCK_INTRA,
-    MPIDI_RANK_MAP_LUT,
-    MPIDI_RANK_MAP_LUT_INTRA,
-    MPIDI_RANK_MAP_MLUT,
-    MPIDI_RANK_MAP_NONE
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/mpid/ch4/include/mpidpre.h:582]: All features remain in the MPIR_Group pmap other than the `MPIDI_RANK_MAP_STRIDE_BLOCK{_INTRA}`, which we simplified away previously in group pmap. We could add that back if want.

[the remark this answers is not in the document]
Comment (raffenet) [reply src/mpid/ch4/include/mpidpre.h:582]: No need to add back. If that's the only difference, I believe it's documented in the commit log already. Just making sure there's no other differences.

File: src/mpid/ch4/src/ch4_comm.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (raffenet) [reply src/mpid/ch4/src/ch4_comm.c:518]: Summarizing offline discussion - this is complex but the hope is that it is isolated enough to allow understanding for anyone who wishes to modify it. The problem is that multiple pieces of the data exchange are dynamically sized, so defining a structured format is challenging. For instance, you can only have one flexible array member at the end of a struct. Multiple messages might offer some more safety, but at the cost of more code. So for now we'll send around this blob 🤷.

File: src/mpid/ch4/netmod/ofi/ofi_spawn.c (hunk 1 above)
Comment (hzhou) [reply src/mpid/ch4/netmod/ofi/ofi_spawn.c:323]: I believe you are right. Added a commit address this and also tried to make it more readable with macro and comments.

File: src/mpid/ch4/src/ch4i_comm.c (hunk 2 above)
Comment (hzhou) [reply src/mpid/ch4/src/ch4i_comm.c:101]: Fixed.

File: src/mpid/ch4/netmod/ofi/ofi_spawn.c (hunk 1 above)
Comment (hzhou) [reply src/mpid/ch4/netmod/ofi/ofi_spawn.c]: Actually `lpid0` should be the formal (non-dynamic) lpid that has an `addr` equal to `0`. So we also need update it if it is not inserted here but being first time copied over from a dynamic entry.

**Review by raffenet — APPROVED** [review]

**Comment by dalcinl** [comment]
@hzhou I suspect this PR caused a regression in mpi4py when building with ch4:ucx:
https://github.com/mpi4py/mpi4py-testing/actions/runs/14184838226/job/39750849021
The build using gforker provides a hint of where to look at:
```
testArgsOnlyAtRoot (test_spawn.TestSpawnMultipleSelf.testArgsOnlyAtRoot) ... Assertion failed in file src/mpid/ch4/src/ch4_proc.c at line 326: world_rank < MPIR_Worlds[world_idx].num_procs
```

PS: When the MPICH codebase goes through large refactors and reimplementations, it would be great if you run the full mpi4py testsuite (at least singleton & 1/5 processes). The multiple regressions in the last months indicate that the MPICH testsuite cannot catch some corner cases.

**Comment by hzhou** [comment]
@dalcinl Are you still having test failures? I tried locally, the tests seem ok.

**Comment by dalcinl** [comment]
> Are you still having test failures?

No, I don't. My CI builds have been green for the last 5 days. I'm wondering whether this was some sort of network hiccup on the GitHub Actions runners. Or did you push a fix that I did not notice?

**Comment by dalcinl** [comment]
Does the assertion output ring a bell? Maybe there is still an issue, but it simply not reproducible and the race condition did not trigger again?

**Comment by hzhou** [comment]
The assertion says the remote process has a rank larger than the their world size. If the dynamic exchange worked, this should not happen. I need a reliable reproducer to track down the bug.
